From 6b5267a384d745c509decc07fe17e994f3a2dad5 Mon Sep 17 00:00:00 2001
From: Karthik M <quic_karm@quicinc.com>
Date: Wed, 28 Feb 2024 16:05:57 +0530
Subject: [PATCH] hostapd: Defer PTK installation for FT auth

Currently for ML STAs, PTK installation happens during FT auth,
which is before adding all link STAs. A ping drop is observed
as PTK is installed only on the assoc link and not on all
other links.

PTK key installation should happens for all the links of station,
hence defer it until all link STAs are added to kernel (i.e after
successful Assoc).

Signed-off-by: Karthik M <quic_karm@quicinc.com>
---
 src/ap/wpa_auth_ft.c | 5 +----
 1 file changed, 1 insertion(+), 4 deletions(-)

--- a/src/ap/wpa_auth_ft.c
+++ b/src/ap/wpa_auth_ft.c
@@ -3318,9 +3318,6 @@ void wpa_ft_install_ptk(struct wpa_state
 		return;
 	}
 
-	if (!retry)
-		wpa_auth_add_sta_ft(sm->wpa_auth, sm->addr);
-
 	/* FIX: add STA entry to kernel/driver here? The set_key will fail
 	 * most likely without this.. At the moment, STA entry is added only
 	 * after association has been completed. This function will be called
@@ -3786,7 +3783,7 @@ pmk_r1_derived:
 	sm->pairwise = pairwise;
 	sm->PTK_valid = true;
 	sm->tk_already_set = false;
-	wpa_ft_install_ptk(sm, 0);
+	wpa_auth_add_sta_ft(sm->wpa_auth, sm->addr);
 
 	if (wpa_ft_set_vlan(sm->wpa_auth, wpa_auth_get_spa(sm), &vlan) < 0) {
 		wpa_printf(MSG_DEBUG, "FT: Failed to configure VLAN");
